docs/design-documents/platform/pin-names/pin_names.md

```diff
@@ -0,0 +1,200 @@
+# Pin Names design document
+
+# Table of contents
+
+1. [Table of contents](#table-of-contents).
+    * [Revision history](#revision-history).
+1. [Introduction](#introduction).
+    1. [Overview and background](#overview-and-background).
+    1. [Requirements and assumptions](#requirements-and-assumptions).
+1. [Detailed design](#detailed-design).
+    1. [Arduino Uno (Rev 3) Header Pins](#arduino-uno-rev-3-header-pins).
+        1. [I2C Definition for Arduino Uno (Rev3) Pins](#i2c-definition-for-arduino-uno-rev3-pins)
+        1. [SPI Definition for Arduino Uno (Rev3) Pins](#spi-definition-for-arduino-uno-rev3-pins)
+    1. [Custom/Vendor Header Pins](#custom/vendor-header-pins).
+        1. [Custom/Vendor Connector I2C Definition](#custom/vendor-connector-i2c-definition).
+        1. [Custom/Vendor Connector SPI Definition](#custom/vendor-connector-spi-definition).
+    1. [Board Components Pins](#board-components-pins).
+        1. [LED Definition](#led-definition)
+        1. [BUTTON Definition](#button-definition)
+1. [Other information](#other-information).
+    1. [Avoid Unnecessary Definitions](#avoid-unnecessary-definitions).
+
+
+### Revision history
+
+1.0 - Initial revision - Malavika Sajikumar - 3/5/2020.  
+This document was written for MBED OS version 5.15 (released on 11th of February 2020).
+
+# Introduction
+
+### Overview and background
+
+Mbed is designed so that application code written in the platform is portable across different Mbed supported boards with the same hardware capabilities or interfaces. However, the code, in most cases, is not truly portable due to the differences in pin name definitions for the same kind of interfaces across different boards. 
+
+This document provides a standard for defining pin names for the commonly used Mbed interfaces like I2C, SPI, etc. and board components like buttons and LEDs in a board support package.
+
+### Requirements and assumptions
+
+This document mainly applies to the pin standards required for Arduino Uno header which is shared among multiple Mbed enabled boards. The Arduino Uno header has been stable since 2012 at its current revision, which is the Arduino Uno Rev3. All design choices in this document for the Arduino Uno header are based on the Arduino Uno Rev3 header implementation.
+
+# Detailed design
+
+To achieve meaningful portability of application code across various Mbed enabled boards, the pin names of commonly used interfaces and board components should be common across these boards. This document describes a set of rules on how to name these pins in the board support package, specifically in PinNames.h file of the board.
+
+### Arduino Uno (Rev3) Header Pins
```

Comment:
We should consider the "Arduino Uno" as one of the popular standards and example of actual implementation, but also need to introduce the concept considering that other standards will follow shortly.

Reply:
> One general comment I have is around the scope of this document. As it is, it addresses the dev boards only. And maybe it is correct, but there's also the production HW and standard pin names (aliases) for the peripherals. Do people think it should live in the HAL spec? In which case we should probably make it obvious that scope of this document is development boards only.
> 
> Second option would be to extend this document about general points which will influence the HAL definition for each peripheral.

@bulislaw The second suggestion here https://github.com/ARMmbed/mbed-os/pull/12838#discussion_r434038918 
> For MCU pins: I2C i2c(ADI::MCU::PB_9 , ADI::MCU::PB_8);
>
can possibly help with easily editing application code for production HW?

Reply:
I believe we need to define the architecture of the pin names on multiple scenarios.
On dev boards, it could be ArduinoUno or other connectors. On production/custom type of boards, the most logical seems to use MCU names with the possibility to let developers customized for their own needs.

I tend to think we should split the document and proposal into two:
- General architecture and guidelines for pin names (do's and don'ts), including MCUs.
- Arduino Uno specific proposal and guidelines, as the first and reference connector

@bulislaw @0xc0170 does it sound reasonable?

Reply:
agreed with @MarceloSalazar